### ISSUE TITLE
Use more rawtypes in package-private implementation types to reduce the jar file size of the runtime library

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#17702]
In the past, there have been a few discussions around ways to reduce the `.jar` file size of jOOQ, which has grown to ~5MB (~14MB of contents, when unpacked) with jOOQ 3.19.15. As we add more features and hardly ever remove any, it obviously grows with every minor release. The discussions have been:

- Reducing feature scope for certain distributions where less scope is OK (rarely useful): https://github.com/jOOQ/jOOQ/issues/8928
- Creating single-dialect specific builds (often useful): https://github.com/jOOQ/jOOQ/issues/13494
- Removing annotations from byte code by changing their retention (rarely problematic, but marginal effect): https://github.com/jOOQ/jOOQ/issues/8868
- Reducing the maximum degree of type safe generic types, e.g. from 22 to 8 (drastically backwards incompatible): https://github.com/jOOQ/jOOQ/issues/9809

Looking again at the top 20 files by file size:

```
$ find . -type f -printf '%s %p\n'|sort -nr|head -n 20
739359 ./impl/DSL.class
391820 ./impl/DefaultDSLContext.class
322663 ./impl/DefaultParseContext.class
248196 ./DSLContext.class
214625 ./impl/Tools.class
190227 ./impl/InsertImpl.class
173839 ./impl/SelectImpl.class
164247 ./impl/MergeImpl.class
163879 ./impl/WithImpl.class
152189 ./xsd/jooq-runtime-3.19.0.xsd
132764 ./conf/Settings.class
130271 ./impl/ConstraintImpl.class
110463 ./Rows.class
109107 ./impl/ResultQueryTrait.class
105232 ./impl/SelectQueryImpl.class
87965 ./impl/UpdateImpl.class
84164 ./WithStep.class
78648 ./impl/ResultImpl.class
69934 ./impl/AbstractField.class
65503 ./impl/RowBetweenCondition.class
```

Let's split them into categories:

**API (or implementation of API convenience methods)**

```
739359 ./impl/DSL.class
391820 ./impl/DefaultDSLContext.class
248196 ./DSLContext.class
190227 ./impl/InsertImpl.class
173839 ./impl/SelectImpl.class
164247 ./impl/MergeImpl.class
163879 ./impl/WithImpl.class
130271 ./impl/ConstraintImpl.class
110463 ./Rows.class
109107 ./impl/ResultQueryTrait.class
87965 ./impl/UpdateImpl.class
84164 ./WithStep.class
78648 ./impl/ResultImpl.class
69934 ./impl/AbstractField.class
65503 ./impl/RowBetweenCondition.class
```

**Implementation (hard to omit)**

```
322663 ./impl/DefaultParseContext.class
214625 ./impl/Tools.class
105232 ./impl/SelectQueryImpl.class
```

**Configuration / data structures**

```
152189 ./xsd/jooq-runtime-3.19.0.xsd
132764 ./conf/Settings.class
```

### The following facts can be established:

- Dialect specific logic is hard to measure with this approach, as it is scattered throughout the codebase. I will run a separate investigation for this at some point, creating a prototype dialect-specific build for only one dialect (e.g. H2): https://github.com/jOOQ/jOOQ/issues/13494. I would imagine, this should also heavily reduce the built jar file and bytecode size.
- API is the most heavy "pollutant" in terms of bytecode size, and not just API in terms of overloads and convenience methods, but generics (!) as shown by the fact how high up trivial classes like `ConstraintImpl` and `Rows` have ranked.

Here's an extract of `ConstraintImpl` byte code:

```
<T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;>(Lorg/jooq/Field<TT1;>;Lorg/jooq/Field<TT2;>;Lorg/jooq/Field<TT3;>;Lorg/jooq/Field<TT4;>;Lorg/jooq/Field<TT5;>;Lorg/jooq/Field<TT6;>;Lorg/jooq/Field<TT7;>;Lorg/jooq/Field<TT8;>;Lorg/jooq/Field<TT9;>;Lorg/jooq/Field<TT10;>;Lorg/jooq/Field<TT11;>;Lorg/jooq/Field<TT12;>;Lorg/jooq/Field<TT13;>;Lorg/jooq/Field<TT14;>;Lorg/jooq/Field<TT15;>;Lorg/jooq/Field<TT16;>;Lorg/jooq/Field<TT17;>;)Lorg/jooq/impl/ConstraintImpl; field18 Lorg/jooq/Field<TT18;>;:<T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;T18:Ljava/lang/Object;>(Lorg/jooq/Field<TT1;>;Lorg/jooq/Field<TT2;>;Lorg/jooq/Field<TT3;>;Lorg/jooq/Field<TT4;>;Lorg/jooq/Field<TT5;>;Lorg/jooq/Field<TT6;>;Lorg/jooq/Field<TT7;>;Lorg/jooq/Field<TT8;>;Lorg/jooq/Field<TT9;>;Lorg/jooq/Field<TT10;>;Lorg/jooq/Field<TT11;>;Lorg/jooq/Field<TT12;>;Lorg/jooq/Field<TT13;>;Lorg/jooq/Field<TT14;>;Lorg/jooq/Field<TT15;>;Lorg/jooq/Field<TT16;>;Lorg/jooq/Field<TT17;>;Lorg/jooq/Field<TT18;>;)Lorg/jooq/impl/ConstraintImpl; field19 Lorg/jooq/Field<TT19;>;g<T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;T18:Ljava/lang/Object;T19:Ljava/lang/Object;>(Lorg/jooq/Field<TT1;>;Lorg/jooq/Field<TT2;>;Lorg/jooq/Field<TT3;>;Lorg/jooq/Field<TT4;>;Lorg/jooq/Field<TT5;>;Lorg/jooq/Field<TT6;>;Lorg/jooq/Field<TT7;>;Lorg/jooq/Field<TT8;>;Lorg/jooq/Field<TT9;>;Lorg/jooq/Field<TT10;>;Lorg/jooq/Field<TT11;>;Lorg/jooq/Field<TT12;>;Lorg/jooq/Field<TT13;>;Lorg/jooq/Field<TT14;>;Lorg/jooq/Field<TT15;>;Lorg/jooq/Field<TT16;>;Lorg/jooq/Field<TT17;>;Lorg/jooq/Field<TT18;>;Lorg/jooq/Field<TT19;>;)Lorg/jooq/impl/ConstraintImpl; field20 Lorg/jooq/Field<TT20;>;”<T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;T18:Ljava/lang/Object;T19:Ljava/lang/Object;T20:Ljava/lang/Object;>(Lorg/jooq/Field<TT1;>;Lorg/jooq/Field<TT2;>;Lorg/jooq/Field<TT3;>;Lorg/jooq/Field<TT4;>;Lorg/jooq/Field<TT5;>;Lorg/jooq/Field<TT6;>;Lorg/jooq/Field<TT7;>;Lorg/jooq/Field<TT8;>;Lorg/jooq/Field<TT9;>;Lorg/jooq/Field<TT10;>;Lorg/jooq/Field<TT11;>;Lorg/jooq/Field<TT12;>;Lorg/jooq/Field<TT13;>;Lorg/jooq/Field<TT14;>;Lorg/jooq/Field<TT15;>;Lorg/jooq/Field<TT16;>;Lorg/jooq/Field<TT17;>;Lorg/jooq/Field<TT18;>;Lorg/jooq/Field<TT19;>;Lorg/jooq/Field<TT20;>;)Lorg/jooq/impl/ConstraintImpl; field21 Lorg/jooq/Field<TT21;>;Á<T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;T18:Ljava/lang/Object;T19:Ljava/lang/Object;T20:Ljava/lang/Object;T21:Ljava/lang/Object;>(Lorg/jooq/Field<TT1;>;Lorg/jooq/Field<TT2;>;Lorg/jooq/Field<TT3;>;Lorg/jooq/Field<TT4;>;Lorg/jooq/Field<TT5;>;Lorg/jooq/Field<TT6;>;Lorg/jooq/Field<TT7;>;Lorg/jooq/Field<TT8;>;Lorg/jooq/Field<TT9;>;Lorg/jooq/Field<TT10;>;Lorg/jooq/Field<TT11;>;Lorg/jooq/Field<TT12;>;Lorg/jooq/Field<TT13;>;Lorg/jooq/Field<TT14;>;Lorg/jooq/Field<TT15;>;Lorg/jooq/Field<TT16;>;Lorg/jooq/Field<TT17;>;Lorg/jooq/Field<TT18;>;Lorg/jooq/Field<TT19;>;Lorg/jooq/Field<TT20;>;Lorg/jooq/Field<TT21;>;)Lorg/jooq/impl/ConstraintImpl; field22 Lorg/jooq/Field<TT22;>;î<T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;T18:Ljava/lang/Object;T19:Ljava/lang/Object;T20:Ljava/lang/Object;T21:Ljava/lang/Object;T22:Ljava/lang/Object;>(Lorg/jooq/Field<TT1;>;Lorg/jooq/Field<TT2;>;Lorg/jooq/Field<TT3;>;Lorg/jooq/Field<TT4;>;Lorg/jooq/Field<TT5;>;Lorg/jooq/Field<TT6;>;Lorg/jooq/Field<TT7;>;Lorg/jooq/Field<TT8;>;Lorg/jooq/Field<TT9;>;Lorg/jooq/Field<TT10;>;Lorg/jooq/Field<TT11;>;Lorg/jooq/Field<TT12;>;Lorg/jooq/Field<TT13;>;Lorg/jooq/Field<TT14;>;Lorg/jooq/Field<TT15;>;Lorg/jooq/Field<TT16;>;Lorg/jooq/Field<TT17;>;Lorg/jooq/Field<TT18;>;Lorg/jooq/Field<TT19;>;Lorg/jooq/Field<TT20;>;Lorg/jooq/Field<TT21;>;Lorg/jooq/Field<TT22;>;)Lorg/jooq/impl/ConstraintImpl; t1 t2 t3 t4 t5 t6 t7 t8 t9 t10 t1
```

And here's `Rows.class`

```
<T:Ljava/lang/Object;T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;T18:Ljava/lang/Object;T19:Ljava/lang/Object;>(Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT1;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT2;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT3;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT4;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT5;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT6;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT7;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT8;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT9;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT10;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT11;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT12;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT13;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT14;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT15;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT16;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT17;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT18;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT19;>;>;)Ljava/util/stream/Collector<TT;*Ljava/util/List<Lorg/jooq/Row19<TT1;TT2;TT3;TT4;TT5;TT6;TT7;TT8;TT9;TT10;TT11;TT12;TT13;TT14;TT15;TT16;TT17;TT18;TT19;>;>;>; f20 ;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT20;>;>;ê<T:Ljava/lang/Object;T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;T18:Ljava/lang/Object;T19:Ljava/lang/Object;T20:Ljava/lang/Object;>(Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT1;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT2;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT3;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT4;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT5;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT6;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT7;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT8;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT9;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT10;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT11;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT12;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT13;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT14;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT15;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT16;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT17;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT18;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT19;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT20;>;>;)Ljava/util/stream/Collector<TT;*[Lorg/jooq/Row20<TT1;TT2;TT3;TT4;TT5;TT6;TT7;TT8;TT9;TT10;TT11;TT12;TT13;TT14;TT15;TT16;TT17;TT18;TT19;TT20;>;>;û<T:Ljava/lang/Object;T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;T18:Ljava/lang/Object;T19:Ljava/lang/Object;T20:Ljava/lang/Object;>(Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT1;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT2;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT3;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT4;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT5;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT6;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT7;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT8;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT9;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT10;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT11;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT12;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT13;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT14;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT15;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT16;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT17;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT18;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT19;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT20;>;>;)Ljava/util/stream/Collector<TT;*Ljava/util/List<Lorg/jooq/Row20<TT1;TT2;TT3;TT4;TT5;TT6;TT7;TT8;TT9;TT10;TT11;TT12;TT13;TT14;TT15;TT16;TT17;TT18;TT19;TT20;>;>;>; f21 ;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT21;>;>;@<T:Ljava/lang/Object;T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;T18:Ljava/lang/Object;T19:Ljava/lang/Object;T20:Ljava/lang/Object;T21:Ljava/lang/Object;>(Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT1;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT2;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT3;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT4;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT5;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT6;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT7;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT8;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT9;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT10;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT11;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT12;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT13;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT14;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT15;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT16;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT17;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT18;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT19;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT20;>;>;Ljava/util/function/Function<-TT;+Lorg/jooq/Field<TT21;>;>;)Ljava/util/stream/Collector<TT;*[Lorg/jooq/Row21<TT1;TT2;TT3;TT4;TT5;TT6;TT7;TT8;TT9;TT10;TT11;TT12;TT13;TT14;TT15;TT16;TT17;TT18;TT19;TT20;TT21;>;>;Q<T:Ljava/lang/Object;T1:Ljava/lang/Object;T2:Ljava/lang/Object;T3:Ljava/lang/Object;T4:Ljava/lang/Object;T5:Ljava/lang/Object;T6:Ljava/lang/Object;T7:Ljava/lang/Object;T8:Ljava/lang/Object;T9:Ljava/lang/Object;T10:Ljava/lang/Object;T11:Ljava/lang/Object;T12:Ljava/lang/Object;T13:Ljava/lang/Object;T14:Ljava/lang/Object;T15:Ljava/lang/Object;T16:Ljava/lang/Object;T17:Ljava/lang/Object;
```

Clearly, we cannot omit generics in `Rows`, which is purely an API type, but perhaps in our implementations, we could reduce the generics overhead where not strictly needed by using rawtypes? Rawtypes aren't going away any time soon. This could be a low hanging fruit.

**Comment by lukaseder** [comment]
On my local machine, where I'm building with JDK 21 by default, the following difference can be established by making only one 22-generic overload implementation in `ConstraintImpl` use rawtypes instead of being correctly generic:

```
Before:
78376 ./org/jooq/impl/ConstraintImpl.class

After:
62959 ./org/jooq/impl/ConstraintImpl.class
```

It's also interesting that JDK 21 produces much smaller `.class` files. Looking at 3.20.0-SNAPSHOT builds, we can see that the `jar` file sizes differ depending on the JDK build:

- JDK 21: 5118003
- JDK 17: 6360114
- JDK 11: 6353166
- JDK 8: 6197575

**Comment by NikolayMetchev** [comment]
Could you make jdk specific versions as well to take advantage of the last fact?

**Comment by lukaseder** [comment]
> Could you make jdk specific versions as well to take advantage of the last fact?

What is *"the last fact*?

**Comment by NikolayMetchev** [comment]
> > Could you make jdk specific versions as well to take advantage of the last fact?
> 
> What is *"the last fact*?
That using JDK 21 produces smaller output

**Comment by lukaseder** [comment]
I don't think that has anything to do with jOOQ. We don't enable any special flags. There's probably a compiler improvement within the JDK itself, or a new class file format.

**Comment by lukaseder** [comment]
Yeah, silly mistake. I compared the jOOQ Open Source Edition for JDK 21 with the other editions for older JDKs. Obviously, it contains less content. I've updated the list now, there's no difference.

But it's a good comparison nonetheless, to hint at how much (or little) will be gained from making dialect specific distributions (https://github.com/jOOQ/jOOQ/issues/13494)

**Comment by lukaseder** [comment]
OK, so I've made the following improvements of various implementation classes, which don't leak into client code.

```
before | after  | file name
---------------------------
 78376 |  62959 | ./org/jooq/impl/ConstraintImpl.class
 66421 |  43221 | ./org/jooq/impl/RowBetweenCondition.class
198937 | 135023 | ./org/jooq/impl/InsertImpl.class
171420 | 135902 | ./org/jooq/impl/WithImpl.class
 88986 |  70658 | ./org/jooq/impl/UpdateImpl.class
 51366 |  33038 | ./org/jooq/impl/DeleteImpl.class
 77586 |  61040 | ./org/jooq/impl/ResultImpl.class
179552 | 148117 | ./org/jooq/impl/MergeImpl.class
 20153 |  18912 | ./org/jooq/impl/RecordImpl22.class
 19266 |  18081 | ./org/jooq/impl/RecordImpl21.class
 18389 |  17260 | ./org/jooq/impl/RecordImpl20.class
 17522 |  16449 | ./org/jooq/impl/RecordImpl19.class
 etc.
 61230 |  48251 | ./org/jooq/impl/RowImpl22.class
 59433 |  46915 | ./org/jooq/impl/RowImpl21.class
 57636 |  45579 | ./org/jooq/impl/RowImpl20.class
 55839 |  44243 | ./org/jooq/impl/RowImpl19.class
 etc.
```

This appears to be a significant enough improvement to go forward with it (still waiting for integration tests on all JDKs. From experience, javac 8 will have a few issues with this improvement).

Regrettably, the jar file size improvement is far less significant, from 6359988 bytes to 6319360 bytes. Unsurprising, because this type of meta data is very easy to compress in a jar file.

I'll keep the topic in the back of my head. Perhaps, a more impressive improvement will become more obvious.

**Comment by lukaseder** [comment]
Implemented in jOOQ 3.20.0

**Comment by lukaseder** [comment]
New 3.20.0-SNAPSHOTS versions have been deployed  with these sizes:

- JDK 21: 6318383
- JDK 17: 6319473
- JDK 11: 6312539
- JDK 8: 6156811

Well, the good news is that at least, we can claim that compression works well :)
